homeassistant/components/cloud/helpers.py

```diff
@@ -0,0 +1,34 @@
+"""Helpers for the cloud component."""
+
+import logging
+import threading
+
+from homeassistant.core import HomeAssistant
+
+
+class FixedSizeQueueLogHandler(logging.Handler):
+    """Log handler to store messages, with auto rotation."""
+
+    MAX_RECORDS = 500
+
+    def __init__(self) -> None:
+        """Initialize a new LogHandler."""
+        super().__init__()
+        self._records: list[logging.LogRecord] = []
```

Comment:
Maybe make this a deque instead since we want to modify both ends of the sequence?

https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-queues

That also gives you the max records feature for free.